src/riscv/lib/src/pvm/linux/signals.rs

```diff
@@ -80,68 +79,50 @@ pub const NO_HANDLER: VirtAddr = VirtAddr::new(0u64);
-        if size_of::<Self>() != STORED_SIZE.get() {
-            panic!("Expected STORED_SIZE to match size_of<LinuxSigAction>");
-        }
-        STORED_SIZE
-    };
+    const STORED_SIZE: NonZeroUsize = { NonZeroUsize::new(SIZE_SIGACTION).unwrap() };
 
     unsafe fn read_unaligned(source: *const u8) -> Self {
+        const SIGRETURN_PADDING: usize = 12;
+
         // SAFETY: The bitwise representation is the same as `write_unaligned` and matches each
         // field.
         unsafe {
             let offset = 0;
 
-            let sa_handler_bits = source.cast::<u64>().read();
-            let offset = offset + size_of_val(&sa_handler_bits);
-
             let sa_sigaction_bits = source.add(offset).cast::<u64>().read();
             let offset = offset + size_of_val(&sa_sigaction_bits);
 
-            let sa_mask_bits = source.add(offset).cast::<u32>().read();
-            let offset = offset + size_of_val(&sa_mask_bits);
-
             let sa_flags_bits = source.add(offset).cast::<u32>().read();
             let offset = offset + size_of_val(&sa_flags_bits);
 
-            let sa_restorer_bits = source.add(offset).cast::<u64>().read();
+            let offset = offset + SIGRETURN_PADDING;
+            let sa_mask_bits = source.add(offset).cast::<u64>().read();
 
             Self {
-                sa_handler: VirtAddr::new(u64::from_le(sa_handler_bits)),
-                sa_sigaction: VirtAddr::new(u64::from_le(sa_sigaction_bits)),
-                sa_mask: u32::from_le(sa_mask_bits),
                 sa_flags: u32::from_le(sa_flags_bits),
-                sa_restorer: VirtAddr::new(u64::from_le(sa_restorer_bits)),
+                sa_sigaction: VirtAddr::new(u64::from_le(sa_sigaction_bits)),
+                sa_mask: u64::from_le(sa_mask_bits),
             }
         }
     }
 
     unsafe fn write_unaligned(self, dest: *mut u8) {
+        const SIGRETURN_PADDING: usize = 12;
         // SAFETY: The bitwise representation is the same as `read_unaligned` and matches each
         // field.
         unsafe {
             let offset = 0;
 
-            dest.cast::<u64>()
-                .write(self.sa_handler.to_machine_address().to_le());
-            let offset = offset + size_of_val(&self.sa_handler);
-
             dest.add(offset)
                 .cast::<u64>()
                 .write(self.sa_sigaction.to_machine_address().to_le());
             let offset = offset + size_of_val(&self.sa_sigaction);
 
-            dest.add(offset).cast::<u32>().write(self.sa_mask.to_le());
-            let offset = offset + size_of_val(&self.sa_mask);
-
             dest.add(offset).cast::<u32>().write(self.sa_flags.to_le());
             let offset = offset + size_of_val(&self.sa_flags);
 
-            dest.add(offset)
-                .cast::<u64>()
-                .write(self.sa_restorer.to_machine_address().to_le());
+            let offset = offset + SIGRETURN_PADDING;
```

Comment:
Is this used to align to `size_of_val(&self.sa_mask)`? Having a function for alignment can be more self-documenting and avoids magic numbers. Maybe adding an `unchecked_align_up` correspondent to the `VirtAddr::align_up` working directly on `u64` ?


```suggestion
            let offset = align_by_increasing(offset, &self.sa_mask);
```

Reply:
>Is this used to align to size_of_val(&self.sa_mask)? 

No, it's because the kernel struct has unused padding between the flags and the mask that would be used by the restorer if the ABI had it as part of the struct. So the padding relates to the restorer, not to the mask.

Would that change things? The value is constant, so it would just be a const function with a similar name returning a constant.